### DISCUSSION TITLE
Enable decorating a service type

### DISCUSSION THREAD

**Discussion by abdulhaq-e** [hynek/svcs#75] [Ideas]
I have a library where I wrote something very similar to `svcs` but not as polished of course :D. I'm planning to remove my hacky dependency resolver code and replace it with `svcs`. However, looking at the source code and examples, there doesn't seem to be a way of "decorating" a service, i.e. registering it once then when registering it a second time, it can have access to the first instance. 

Am I right with this conclusion? I can attempt to open a PR with my solution if that is the case.

**Comment by abdulhaq-e** [comment]
```python

import svcs
import uuid
from typing import Protocol

class UserService(Protocol):
    pass

class UserSQLAlchemyService:
   pass

registry = svcs.Registry()

registry.register_factory(UserService, lambda: UserSQLAlchemyService)

class UserSQLAlchemyAdvancedService:
   def __init__(self, service: UserService):
        self.service = service

def user_sqlalchemy_service_decorator(svcs_container):
    original_service = svcs_container.get(UserService)
    return UserSQLAlchemyAdvancedService(original_service)

registry.register_factory(UserService, user_sqlalchemy_service_decorator)
```

Something like the above, I'm relying on a plugin architecture where decorating existing services is important. In my implementation, the the decoration has a special api and they are all stored in an iterable where they are applied after all "main" registrations are resolved.

So something like this would be possible:
```python
registry.register_factory(UserService, user_sqlalchemy_service_decorator, decorator=True)
```
or to be more explicit:
```python
registry.decorate_factory(UserService, user_sqlalchemy_service_decorator)
```

**Comment by hynek** [comment]
hey I'm super sorry for the delays – this is literally the oldest notification in my inbox and I've started replying, trying to understand, and abandoning for "later" a dozen times.

Generally speaking, and of course only if you still care, am I understanding correctly, that you want to be a) able to overwrite type registrations while b) accessing the old registration _at runtime_? This sounds like a severe foot-gun to me. Is anything really stopping you from using bespoke types for your intermediate states?

**Comment by abdulhaq-e** [comment]
Hello hynek, thank you for getting back to me. I completely I understand, in fact, I consider myself lucky to have seen this notification only after 5 days. Github notifications are so noisy and I've been planning to filter them for years!

While I agree this is a severe foot gun for an application code since it introduces implicit behaviour, I'd say it's inevitable for a library. In a library with a plugin system, the custom types are not known beforehand therefore they must rely on decorating a unified type if needed. An example would roughly look like this:

main library code:
```python
class App(Protocol):
  def __call__(self, request: Request) -> Response: ...

def core_dispatch(r: Request) -> Response:
  return Response()
        
registry.register(App, lambda: core_dispatch)
```

then a third party plugin needs to decorate how the `App` works, the plugin has access to the container/registry:

```python
class GzipPlugin:
  def __call__(self, registry):
    def decorate(container):
        inner_app = container.get(App)
        return lambda req: gzip_response(inner_app(req))
    registry.decorate(App, decorate)
```

here is my messy implementation: https://github.com/mersal-org/mersal/blob/main/mersal/configuration/dependency_resolver.py
